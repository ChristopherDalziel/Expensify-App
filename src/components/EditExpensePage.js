import React from "react";

const EditExpensePage = () => {
  return <div>Edit Expense Page</div>;
};

export default EditExpensePage;
