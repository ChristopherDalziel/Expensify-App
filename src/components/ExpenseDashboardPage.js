import React from "react";

const ExpenseDashboardPage = () => {
  return <div>hello</div>;
};

export default ExpenseDashboardPage;
