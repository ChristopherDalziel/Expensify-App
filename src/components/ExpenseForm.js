import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import styled from "@emotion/styled";

// const date = new Date();
// const now = moment();

// console.log(now.format("MMM Do YYYY"));

const FormContainer = styled.div`
  margin: 10px;
  width: 50vw;
`;

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      // On the amount state, if we're passing it back from an expense we need to convert it back from a number to a string
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      // The createdAt state needs to take in the timestamp of the expense not the current time.
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => {
      return { description };
    });
  };

  onTextAreChange = (e) => {
    const note = e.target.value;
    this.setState(() => {
      return { note };
    });
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    // Regular expressions, setting restrictions on what we will allow to be set in our state (Checkout regex101.com)
    // If there is no amount or if the amount matches set a new state (This allows the user to remove their input, without this you cannot.)
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return { amount };
      });
    }
  };

  onDateChange = (createdAt) => {
    // If a there is a createdAt value, use this.setState, if there is not.. do nothing. This prevents the user from removing the date entirely.
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide a description and an amount",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        // We need to convert amount into the correct type, by converting it from a string. Float keeps the decmimal places, and setting the 10 afterwards ensures we're in base 10. Since we're working in cents we need to multiply by 100.
        amount: parseFloat(this.state.amount, 10) * 100,
        // Value of is a moment function and allows us to format the time/date correctly
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <FormContainer>
        <Form onSubmit={this.onSubmit}>
          {this.state.error && <p>{this.state.error}</p>}
          <Form.Label>Expense Description</Form.Label>
          <Form.Control
            autoFocus
            type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          ></Form.Control>
          <Form.Label>Expense Amount</Form.Label>
          <Form.Control
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          ></Form.Control>
          <Form.Label>Expense Date</Form.Label>
          <br />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <br />
          <Form.Label>Expense Notes</Form.Label>
          <Form.Control
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onTextAreChange}
          ></Form.Control>
          <br />
          <Button variant="outline-primary" type="submit">
            Add Expense
          </Button>
        </Form>
      </FormContainer>
    );
  }
}
