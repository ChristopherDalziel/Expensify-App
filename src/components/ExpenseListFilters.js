import React from "react";
import { connect } from "react-redux";
import { setTextFilter } from "../actions/filters";

const ExpenseListFilters = (props) => (
  <div>
    {/* This input is reading and writing from the store! */}
    <input
      type="text"
      value={props.filters.text}
      onChange={(e) => {
        // Attaching our input to our action, dispatch will change the value in our store.
        props.dispatch(setTextFilter(e.target.value));
      }}
    ></input>
  </div>
);

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
