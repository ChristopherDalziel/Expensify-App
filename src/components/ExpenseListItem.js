import React from "react";
import { connect } from "react-redux";
import { removeExpense } from "../actions/expenses";

const ExpenseListItem = ({ dispatch, description, amount, createdAt, id }) => (
  <div>
    <h3>{description}</h3>
    <p>
      {amount} - {createdAt}
    </p>
    <button
      onClick={(e) => {
        dispatch(removeExpense({ id }));
      }}
    >
      Remove
    </button>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
  };
};

export default connect()(ExpenseListItem);
