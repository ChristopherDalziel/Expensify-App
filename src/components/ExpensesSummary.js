import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import getVisibleExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expensesTotal";
import styled from "@emotion/styled";

const ExpensesSummaryContainer = styled.div`
  margin: 10px;
`;

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? "expense" : "expenses";
  const formattedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00");
  return (
    <ExpensesSummaryContainer>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
      </h1>
    </ExpensesSummaryContainer>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
