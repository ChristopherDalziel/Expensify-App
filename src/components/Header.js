import React from "react";
import { NavLink } from "react-router-dom";

import "../styles/header.css";

const Header = () => {
  return (
    <div>
      <h1>Expensify</h1>
      <NavLink exact activeClassName="isActive" to="/">
        Dashboard
      </NavLink>
      <NavLink activeClassName="isActive" to="/create">
        Create
      </NavLink>
    </div>
  );
};

export default Header;
