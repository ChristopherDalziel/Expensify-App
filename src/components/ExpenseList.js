import React from "react";
import { connect } from "react-redux";

import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

// Regular Component
export const ExpenseList = (props) => (
  <div>
    {props.expenses.length === 0 ? (
      <p>No expenses to display</p>
    ) : (
      props.expenses.map((expense) => {
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);

// Function
const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

// Higher order component
// export default connect((state) => {
//   return {
//     expenses: state.expenses,
//   };
// })(ExpenseList);

// Connect call that pulls our data together, the result of this is a brand new component which is our component with the props from the store
export default connect(mapStateToProps)(ExpenseList);
