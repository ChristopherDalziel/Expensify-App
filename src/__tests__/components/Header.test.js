import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { Header } from "../../components/Header";

test("Should render header correctly", () => {
  const wrapper = shallow(<Header startLogout={() => {}} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("Should call startLogout on button click", () => {
  const startLogoutSpy = jest.fn();
  const wrapper = shallow(<Header startLogout={startLogoutSpy} />);
  wrapper.find("NavLink").at(2).simulate("click");
  expect(startLogoutSpy).toHaveBeenCalledTimes(1);
});
