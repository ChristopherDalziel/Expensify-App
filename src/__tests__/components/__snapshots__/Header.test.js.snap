// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render header correctly 1`] = `
<Navbar
  collapseOnSelect={false}
  expand={true}
  variant="light"
>
  <h1>
    Expensify
  </h1>
  <NavItem>
    <NavLink
      activeClassName="isActive"
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "NavLink",
          "propTypes": Object {
            "activeClassName": [Function],
            "activeStyle": [Function],
            "aria-current": [Function],
            "className": [Function],
            "exact": [Function],
            "innerRef": [Function],
            "isActive": [Function],
            "location": [Function],
            "onClick": [Function],
            "replace": [Function],
            "strict": [Function],
            "style": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      disabled={false}
      exact={true}
      to="/dashboard"
    >
      Dashboard
    </NavLink>
  </NavItem>
  <NavItem>
    <NavLink
      activeClassName="isActive"
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "NavLink",
          "propTypes": Object {
            "activeClassName": [Function],
            "activeStyle": [Function],
            "aria-current": [Function],
            "className": [Function],
            "exact": [Function],
            "innerRef": [Function],
            "isActive": [Function],
            "location": [Function],
            "onClick": [Function],
            "replace": [Function],
            "strict": [Function],
            "style": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      disabled={false}
      to="/create"
    >
      Create
    </NavLink>
  </NavItem>
  <NavLink
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "SafeAnchor",
        "render": [Function],
      }
    }
    disabled={false}
    onClick={[Function]}
  >
    Logout
  </NavLink>
</Navbar>
`;
